data-processing/pdf/process_pdf.py

```diff
@@ -0,0 +1,236 @@
+import requests
+import logging
+from typing import List, Dict
+import sys
+from tempfile import TemporaryDirectory
+import subprocess
+import time
+
+import pdf.grobid_client
+from common.models import BoundingBox as BoundingBoxModel
+from common.models import init_database_connections
+from common.types import CitationData, CitationLocation, SerializableReference, BoundingBox, Matches, Match, Symbol
+from entities.symbols.types import SymbolData, SymbolWithId, SymbolId
+from entities.citations.utils import upload_citations, extract_ngrams, ngram_sim
+from entities.symbols.utils import upload_symbols
+
+class PdfStructureParser(object):
+    def __init__(self, pdf_hash, structure_map):
+        self.pdf_hash = pdf_hash
+        self.structure_map = structure_map
+        self.pages = structure_map['tokens']['pages']
+        self.page_sizes = dict((p['page']['pageNumber'], p['page']) for p in self.pages)
+        self.page_indices = dict((p['page']['pageNumber'], i) for i,p in enumerate(self.pages))
+        self.elements = structure_map['elements']['elementTypes']
+        self.references = None
+
+    def find_token(self, index):
+        for p in self.pages:
+            if (index < len(p['tokens'])):
+                return p['page'], p['tokens'][index]
+            else:
+                index -= len(p['tokens'])
+        return None
+
+
+    def get_text(self,spans):
+        text = []
+        for s in spans:
+            direct = s.get('dehyphenizedText')
+            if direct:
+                text += direct
+            else:
+                for i in range(s['left'], s['right']):
+                    page, token = self.find_token(i)
+                    text.append(token['text'])
+        return ' '.join(text)
+
+    def union(self, bbox1: BoundingBoxModel, bbox2: BoundingBoxModel):
+        if not bbox1:
+            return bbox2
+        x1 = min(bbox1.left,bbox2.left)
```

Comment:
Just double-checking as I know that has bitten in the butt more than once with PDF: `top` = the lower number (i.e. measured from the top of the document) and `bottom` = the higher number?